AMDGPU: Implement tensor load and store instructions for gfx1250

changpeng opened the pull request.

changpeng committed:
AMDGPU: Implement tensor load and store instructions for gfx1250

Co-Authored-by: Stanislav Mekhanoshin <Stanislav.Mekhanoshin@amd.com>

arsenm reviewed (commented):
Fix clang-format error

tgymnich commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -5354,6 +5368,22 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
+    case Intrinsic::amdgcn_tensor_store_from_lds: {
+      // Lie and claim everything is legal, even all operands need to be
+      // SGPRs. applyMapping will have to deal with it with readfirstlane.
+      for (unsigned I = 1; I < MI.getNumOperands(); ++I) {
+        if (MI.getOperand(I).isReg()) {
+          Register Reg = MI.getOperand(I).getReg();
+          auto OpBank = getRegBankID(Reg, MRI);
+          unsigned Size = getSizeInBits(Reg, MRI, *TRI);
+          OpdsMapping[I] = AMDGPU::getValueMapping(OpBank, Size);
+        }
+      }
```
```suggestion
      for (Register Reg : MI.all_uses()) {
        auto OpBank = getRegBankID(Reg, MRI);
        unsigned Size = getSizeInBits(Reg, MRI, *TRI);
        OpdsMapping[I] = AMDGPU::getValueMapping(OpBank, Size);
      }
```

tgymnich replied:
sorry. Updated the code accordingly.

changpeng replied:
> sorry. Updated the code accordingly.

What is the value of "I" then?  We need a way to map the operand back to the index.

jmmartinez commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -3580,6 +3580,37 @@ def int_amdgcn_fdiv_fast : DefaultAttrsIntrinsic<
+    "", [SDNPMemOperand]
+  >;
+
+def int_amdgcn_tensor_load_to_lds : AMDGPUTensorLoadStore;
```
Can we use `ClangBuiltin` in here to avoid the boilerplate in `TargetBuiltins/AMDGPU.cpp`?

```
def int_amdgcn_tensor_load_to_lds : ClangBuiltin<"__builtin_amdgcn_tensor_load_to_lds">, AMDGPUTensorLoadStore;
```

changpeng replied:
This is great. Thanks for the suggestion and explanation

changpeng committed:
AMDGPU: Implement tensor load and store instructions for gfx1250

 A few update based on review comments:

1. fix clang format
2. use ClangBuiltin in intrinsic defs to avoid the boilerplate in
   TargetBuiltins/AMDGPU.cpp
3. Fold to return of bool expression

changpeng commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -621,6 +621,32 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
+
+    SmallVector<Value *, 5> Args;
+    for (int i = 0, e = E->getNumArgs(); i != e; ++i)
+      Args.push_back(EmitScalarExpr(E->getArg(i)));
```
No longer relevant after Using  ClangBuiltin in intrinsic definitions.

changpeng committed:
AMDGPU: Implement tensor load and store instructions for gfx1250

  Minor format change.
